Add WWH OBD DTC message handling

jacobs-rvw opened the pull request:
Add support for 0x42 reportWWHOBDDTCByMaskRecord.

Closes: https://github.com/pylessard/python-udsoncan/issues/245

pylessard commented on udsoncan/services/ReadDTCInformation.py:
```diff
@@ -872,6 +877,33 @@ def interpret_response(cls,
+            _dtc_status_availability_mask = response.data[2]
+            _dtc_severity_availability_mask = response.data[3]
+            _dtc_format_identifier = response.data[4]
+            print(_functional_group_id, _dtc_status_availability_mask, _dtc_severity_availability_mask, _dtc_format_identifier)
```
Debug print?

pylessard commented on udsoncan/services/ReadDTCInformation.py:
```diff
@@ -872,6 +877,33 @@ def interpret_response(cls,
+            print(_functional_group_id, _dtc_status_availability_mask, _dtc_severity_availability_mask, _dtc_format_identifier)
+
+            remaining_bytes = response.data[5:]
+            print([hex(x) for x in remaining_bytes])
```
Debug print?

jacobs-rvw replied:
Removed those   - thanks

jacobs-rvw commented:
If possible sometime later this week. I'm checking with the team to see if there were any other issues we're looking to resolve but for the time being this is the main one. I'll check the CI results in a bit. It didn't seem immediately obvious what the failure was.

pylessard commented:
just restarted ci.  venv creation failed.. I don't think it's coming from you.

pylessard commented:
Possibly this : https://github.com/pylessard/python-udsoncan/commit/9ff5f4a11dc0620a98f9d0fbd2d1e754efdb07fa

I think my master branch was passing because it was reusing a venv created previously. 
I fix the master branch. You can merge or rebase your PR.

Don't forget to remove the debug prints

pylessard commented:
I will leave this here. It's a change made in setuptools. 

https://github.com/apache/spark/pull/43891
https://github.com/apache/spark/pull/50371

sidd91 commented on udsoncan/client.py:
```diff
@@ -1361,6 +1361,27 @@ def get_dtc_by_status_severity_mask(self, status_mask: int, severity_mask: int)
         """
         return self.read_dtc_information(services.ReadDTCInformation.Subfunction.reportDTCBySeverityMaskRecord, status_mask=status_mask, severity_mask=severity_mask)
 
+    def get_wwh_obd_dtc_by_status_mask(self, functional_group_id: int, dtc_class: int, status_mask: int, severity_mask: int) -> Optional[services.ReadDTCInformation.InterpretedResponse]:
```
It will be a cleaner interface if the things such as functional_group_id, dtc_class etc are defined as enum.

pylessard replied:
I haven't reviewed the standard yet, but module is shy on enum on purpose. Most (all?) functions expect integer, but there are constants defined with known values. Example : https://github.com/pylessard/python-udsoncan/blob/master/udsoncan/services/SecurityAccess.py#L24-L52


The rationale is that the standard often defines few values and leaves the other as "ManufacturerSpecific". If we use strongly typed enum, it will start becoming painful to use.

jacobs-rvw replied:
I clarified in the docstrings how they're intended to be used to avoid confusion since the spec wasn't great about isolating the two fields.

pylessard commented on udsoncan/services/ReadDTCInformation.py:
```diff
@@ -361,6 +363,9 @@ def make_request(cls,
             req.data = struct.pack('B', extended_data_record_number)
 
+        elif subfunction == ReadDTCInformation.Subfunction.reportWWHOBDDTCByMaskRecord:
+            req.data = struct.pack('BBB', functional_group_id, status_mask, severity_mask)
```
please check for functional_id range. 00 to FE (inclusively). 0xFF is invalid

jacobs-rvw replied:
I added the asserts using the existing two functions. I can make a new one for functional group if you want - not sure if the others use it.

pylessard replied:
I don't have a strong opinion. As long as it is checked and code make sense

pylessard replied:
Looking at this with more focus. I would prefer a check that returns an error message. The assert functions i wrote are meant to throw an exception with a meaningful error string. This module is meant to guide the programmer to avoid all the little caveat of the standard. Having a plain assert that fails might make some people scratch their head.

pylessard commented on udsoncan/client.py:
```diff
@@ -1361,6 +1361,27 @@ def get_dtc_by_status_severity_mask(self, status_mask: int, severity_mask: int)
+
+        :Effective configuration: ``exception_on_<type>_response`` ``tolerate_zero_padding`` ``ignore_all_zero_dtc``
+
+        :param status_mask: The status mask against which the DTCs are tested. 
```
Missing 2 params in docstring

jacobs-rvw replied:
I just updated these - that's true thanks

pylessard commented:
Can you add constants to define the common values of ``FunctionalGroupIdentifiers``. There seems to be only 3 value defined.

Emissions-system group : 0x33
Safety-system group: 0xD0
VOBD system: 0xFE

You can add a class inside common/dtc.py So we can have ``Dtc.FunctionalGroupIdentifiers``. Inside your method docstring, you can mention that the values can comes from ``:class:`Dtc.FunctionalGroupIdentifiers<Dtc.FunctionalGroupIdentifiers>` ``

Finally, add an autoclass  in doc/source/udsoncan/helper_classes.rst.  Sphinx will connect your docstring to it automatically.

jacobs-rvw commented:
I updated the docs. Let me know if there's anything else. I'd like to squash and reword once you're happy. I need to validate a few things tomorrow but the unit tests look ok to me and I think the ergonomics are improved.

Thanks for the help.

pylessard commented:
Will do a final check later.
I only do squash-merge btw, so do 't worry too much about the history

pylessard commented on udsoncan/common/dtc.py:
```diff
@@ -37,6 +37,14 @@ def get_name(cls, given_id: Optional[int]) -> Optional[str]:
 
+    class FunctionalGroupIdentifiers:
+        """
+        Provides a list of FunctionalGroupIdentifiers (Table D.15) which are used by the :ref:`The ReadDTCInformation<ReadDtcInformation>` when requesting a number of DTCs.
```
typo: "the the".  missing "service" word

pylessard commented:
You have my 2 final comments.
Call the shot and I will merge.
Let me know iwhen you need the release. It takes less than a minute to make

Krash-2504 commented:
@jacobs-rvw If possible could you add support for report wwhobd dtc with permanent status?

jacobs-rvw commented:
@pylessard - I resolved those two findings, squashed and cleaned up. Tested again and it seems to meet the our needs. If you can merge/release this week it would be appreciated. Thanks for the help.
